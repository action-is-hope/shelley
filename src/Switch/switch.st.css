@st-namespace "Switch";

@st-import InputLabel from "../Label/label.st.css";

.root {
  -st-states:
    isDisabled,
    isFocusVisible,
    size(number);
}

.inputLabel {
  -st-extends: InputLabel;
}

.inputContainer {}

.input {}
