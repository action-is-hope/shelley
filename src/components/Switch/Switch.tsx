import React, { forwardRef, useRef } from "react";
import Label from "../Label/Label";
import { useSwitch } from "@react-aria/switch";
import { useToggleState } from "@react-stately/toggle";
import { mergeRefs, mergeProps } from "@react-aria/utils";
import { useFocusRing } from "react-aria";
import type { AriaSwitchProps } from "@react-types/switch";
import type { Size, AlignPos, ComponentBase } from "../types";
import { st, classes } from "./switch.st.css";

export interface SwitchProps extends AriaSwitchProps, ComponentBase {
  className?: string;
  /** The position of the label relative to the label. */
  inputPosition?: AlignPos;
  /** Visually hide the label so it is still accessible to assistive technologies. */
  visuallyHideLabel?: boolean;
  /** Size of the actual input. */
  size?: Size;
}

function Switch(props: SwitchProps, ref: React.Ref<HTMLInputElement>) {
  const {
    className: classNameProp,
    children,
    visuallyHideLabel,
    inputPosition,
    size = 1,
    isDisabled,
    "data-id": dataId,
  } = props;

  const localRef = useRef(null);

  const state = useToggleState(props);
  const { inputProps } = useSwitch(props, state, localRef);

  const { isFocusVisible, focusProps } = useFocusRing();

  const classNames = st(
    classes.root,
    {
      isDisabled,
      isFocusVisible,
      size: size ? size : undefined,
    },
    classNameProp
  );

  const inputControl = (
    <span className={classes.inputContainer}>
      <input
        className={classes.input}
        {...mergeProps(inputProps, focusProps)}
        ref={ref ? mergeRefs(ref, localRef) : localRef}
        data-id={dataId ? `${dataId}--input` : undefined}
      />
    </span>
  );

  return (
    <Label
      className={classNames}
      {...{ inputControl, inputPosition }}
      visuallyHidden={visuallyHideLabel}
      data-id={dataId ? `${dataId}--label` : undefined}
    >
      {children}
    </Label>
  );
}

const _Switch = forwardRef(Switch);
export { _Switch as Switch };
