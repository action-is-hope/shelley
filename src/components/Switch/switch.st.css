@namespace "Switch";

@st-import InputLabel from "../../components/Label/label.st.css";

.root {
  -st-states:
    isDisabled,
    isFocusVisible,
    size(number);
}

.inputLabel {
  -st-extends: InputLabel;
}

.inputContainer {}

.input {}
