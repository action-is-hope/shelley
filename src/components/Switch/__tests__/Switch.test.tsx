import { Switch } from "../Switch";
import renderer from "react-test-renderer";

describe("Switch", () => {
  it("renders correctly with label", () => {
    const tree = renderer.create(<Switch>Label</Switch>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with custom label", () => {
    const tree = renderer
      .create(
        <>
          <label htmlFor="test123"></label>
          <Switch id="test123" />
        </>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
