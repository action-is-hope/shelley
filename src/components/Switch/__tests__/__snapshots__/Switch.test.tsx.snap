// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders correctly with custom label 1`] = `
[
  <label
    htmlFor="test123"
  />,
  <span
    className="Switch__Test__root Switch__Test---vol-1-1"
  >
    <span
      className="Switch__Test__inputContainer"
    >
      <input
        checked={false}
        className="Switch__Test__input"
        disabled={false}
        id="test123"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="switch"
        type="checkbox"
      />
    </span>
  </span>,
]
`;

exports[`Switch renders correctly with label 1`] = `
<label
  className="Label__Test__root Label__Test--hasInput Switch__Test__root Switch__Test---vol-1-1"
>
  <span
    className="Label__Test__labelText"
  >
    Label
  </span>
  <span
    className="Switch__Test__inputContainer"
  >
    <input
      checked={false}
      className="Switch__Test__input"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="switch"
      type="checkbox"
    />
  </span>
</label>
`;
